modulemd/modulemd-module-stream.c

```diff
@@ -651,6 +690,704 @@ modulemd_module_stream_upgrade_to_v2 (ModulemdModuleStream *from)
+      build_config_add_requirement =
+        &modulemd_build_config_add_runtime_requirement;
+    }
+
+  g_debug ("Expansion: stream_expansion_helper (%s) called", which);
+
+  /* for each module... */
+  for (guint i = 0; i < g_strv_length (module_list); i++)
+    {
+      module = module_list[i];
+      streams = (*dependencies_get_streams_as_strv) (deps, module);
+
+      g_debug ("Expansion: module %s dependency %s has %d streams",
+               which,
+               module,
+               g_strv_length (streams));
+
+      if (g_strv_length (streams) == 0)
+        {
+          g_set_error (error,
+                       MODULEMD_ERROR,
+                       MMD_ERROR_UPGRADE,
+                       "Cannot expand module %s dependency %s for all active "
+                       "existing streams.",
+                       which,
+                       module);
+          return FALSE;
+        }
+
+      new_expanded_deps = g_ptr_array_new_with_free_func (g_object_unref);
+
+      /* for each module stream... */
+      for (guint j = 0; j < g_strv_length (streams); j++)
+        {
+          stream = streams[j];
+          g_debug ("Expansion: looking at %s stream dependency %s:%s",
+                   which,
+                   module,
+                   stream);
+
+          if (stream[0] == '-')
+            {
+              g_set_error (error,
+                           MODULEMD_ERROR,
+                           MMD_ERROR_UPGRADE,
+                           "Cannot expand module %s dependency %s using "
+                           "stream exclusion (%s).",
+                           which,
+                           module,
+                           stream);
+              return FALSE;
+            }
+
+          /* if no expanded dependencies yet, just create a new dep for this module and stream */
+          if ((*expanded_deps)->len == 0)
+            {
+              g_debug ("Expansion: creating new dependency");
+
+              new_dep = modulemd_build_config_new ();
+              (*build_config_add_requirement) (new_dep, module, stream);
+              g_ptr_array_add (new_expanded_deps, g_steal_pointer (&new_dep));
+            }
+          /* otherwise, expand on what we already have */
+          else
+            {
+              /* for every expanded dependency we have so far... */
+              for (guint k = 0; k < (*expanded_deps)->len; k++)
+                {
+                  g_debug ("Expansion: expanding existing dependency");
+                  /* Make a copy of the existing expanded dependency and add this module and stream */
+                  new_dep = modulemd_build_config_copy (
+                    g_ptr_array_index (*expanded_deps, k));
+                  (*build_config_add_requirement) (new_dep, module, stream);
+                  g_ptr_array_add (new_expanded_deps,
+                                   g_steal_pointer (&new_dep));
+                }
+            }
```

Comment:
This loop is the very heart of the stream expansion. It's iterative expanding the existing dependency set (`expanded_deps`) by making a copy of each item, adding the current module:stream as a dependency, then adding that updated copy to `new_expanded_deps`. `expanded_deps` stays unmodified until the end of the module loop when all the streams have been added as new dependences. Then it gets cleared and replaced by `new_expanded_deps`.  I'm sure I _could_ use `g_ptr_array_extend_and_steal()`, but I think it would mean creating yet another `GPtrArray` array, adding things to that, and then extending/stealing. I hate to make it even more difficult to understand!

Reply:
OK, I clearly didn't understand the algorithm well enough here. I think that's a good sign that it needs to have some heavy commenting done to explain things.

Reply:
Yeah. Stream expansion starts by generating a full Cartesian product of the StreamV2 dependencies to "flatten" them so they can be mapped to StreamV3. This is the guts of calculating that product and storing it in an array of Buildconfig objects. I'll work on adding commentary text.

Reply:
I just pushed a commit adding a lot of commentary. Hopefully it helps in understanding the code. (And it was good for me to try to explain it!)